src/StmtToViz.cpp

```diff
@@ -0,0 +1,2900 @@
+    }
+
+    void visit(const Fork *op) override {
+        IRVisitor::visit(op);
+        set_compute_costs(op, 0, {op->first.get(), op->rest.get()}, {});
+        set_data_costs(op, 0, {op->first.get(), op->rest.get()}, {});
+    }
+
+    void visit(const IfThenElse *op) override {
+        IRVisitor::visit(op);
+        set_compute_costs(op, 1, {op->condition.get(), op->then_case.get()}, {op->condition.get()});
+        set_data_costs(op, 0, {op->condition.get(), op->then_case.get()}, {op->condition.get()});
+    }
+
+    void visit(const Evaluate *op) override {
+        IRVisitor::visit(op);
+        set_compute_costs(op, 0, {op->value.get()});
+        set_data_costs(op, 0, {op->value.get()});
+    }
+
+    void visit(const Atomic *op) override {
+        IRVisitor::visit(op);
+        set_compute_costs(op, 0, {op->body.get()}, {});
+        set_data_costs(op, 0, {op->body.get()}, {});
+    }
+};
+
+/******************* GetAssemblyInfo *******************/
+// Used to map some Halide IR nodes to line-numbers in the
+// assembly file containing the corresponding generated code.
+class AssemblyInfo : public IRVisitor {
+public:
+    AssemblyInfo()
+        : _loop_id(0), _prodcons_id(0) {
+    }
+
+    void generate(const string &code, const Module &m) {
+        // Traverse the module to populate the list of
+        // nodes we need to map and generate their assembly
+        // markers (comments that appear in the assembly code
+        // associating the code with this node)
+        for (const auto &fn : m.functions()) {
+            fn.body.accept(this);
+        }
+
+        // Find markers in asm code
+        istringstream _asm_stream(code);
+        string line;
+        int lno = 1;
+        while (getline(_asm_stream, line)) {
+            // Try all markers
+            std::vector<uint64_t> matched_nodes;
+            for (auto const &[node, regex] : _markers) {
+                if (std::regex_search(line, regex)) {
+                    // Save line number
+                    _lnos[node] = lno;
+                    // Save this node's id
+                    matched_nodes.push_back(node);
+                }
+            }
+            // We map to the first match, stop
+            // checking matched nodes
+            for (auto const &node : matched_nodes) {
+                _markers.erase(node);
+            }
+
+            lno++;
+        }
+    }
+
+    int get_asm_lno(uint64_t node_id) {
+        if (_lnos.count(node_id)) {
+            return _lnos[node_id];
+        }
+        return -1;
+    }
+
+private:
+    // Generate asm markers for Halide loops
+    int _loop_id;
+    int gen_loop_id() {
+        return ++_loop_id;
+    }
+
+    std::regex gen_loop_asm_marker(int id, const string &loop_var) {
+        std::regex dollar("\\$");
+        string marker = "%\"" + std::to_string(id) + "_for " + loop_var;
+        marker = std::regex_replace(marker, dollar, "\\$");
```

Comment:
Does this need a regex to do the substitution? If not using `replace_all` (from Util.h) is likely to be cheaper and simpler to maintain.

EDIT: oh, I see, you're saving the regex for later. Yeah, I guess this works.